R/tdata.R

```diff
@@ -0,0 +1,188 @@
+#' )
+#'
+#' # Extract a data.frame
+#' isolate(data[["iris"]]())
+#'
+#' # Get code
+#' isolate(get_code_tdata(data))
+#'
+#' # Get metadata
+#' get_metadata(data, "iris")
+#'
+#' @export
+new_tdata <- function(data, code = "", join_keys = NULL, metadata = NULL, check = FALSE, label) {
+  checkmate::assert_list(
+    data,
+    any.missing = FALSE, names = "unique",
+    types = c("data.frame", "reactive", "MultiAssayExperiment")
+  )
+  checkmate::assert_class(join_keys, "JoinKeys", null.ok = TRUE)
+  checkmate::assert_multi_class(code, c("character", "reactive", "CodeClass"))
+
+  checkmate::assert_list(metadata, names = "unique", null.ok = TRUE)
+  checkmate::assert_subset(names(metadata), names(data))
+  for (m in metadata) teal.data::validate_metadata(m)
+
+  if (is.reactive(code)) {
+    isolate(checkmate::assert_multi_class(code(), c("character", "CodeClass"), .var.name = "code"))
+  }
+
+  # create reactive data.frames
+  for (x in names(data)) {
+    if (!is.reactive(data[[x]])) {
+      data[[x]] <- do.call(reactive, list(as.name(x)), envir = list2env(data[x]))
+    } else {
+      isolate(
+        checkmate::assert_multi_class(
+          data[[x]](), c("data.frame", "MultiAssayExperiment"),
+          .var.name = "data"
+        )
+      )
+    }
+  }
+
+  # code #nolint
+  # code <- if (is.reactive(code) && is.character(code())) { #nolint
+  #   CodeClass$new(code()) #nolint
+  # } else if (is.character(code)) { #nolint
+  #   CodeClass$new(code) #nolint
+  # } #nolint
+
+  # set attributes
+  attr(data, "code") <- if (is.reactive(code)) code else reactive(code)
+  attr(data, "join_keys") <- join_keys
+  attr(data, "metadata") <- metadata
+  attr(data, "check") <- check
+  attr(data, "label") <- label
+
+  # set class
+  class(data) <- c("tdata", class(data))
+  data
+}
+
+#' Function to convert a `tdata` object to an `environment`
+#' Any `reactives` inside `tdata` are first evaluated
+#' @param data a `tdata` object
+#' @return an `environment`
+#' @examples
+#'
+#' data <- new_tdata(
+#'   data = list(iris = iris, mtcars = reactive(mtcars)),
+#'   code = "iris <- iris
+#'     mtcars = mtcars",
+#'   label = list(iris = "iris", mtcars = "mtcars")
+#' )
+#'
+#' my_env <- isolate(tdata2env(data))
+#'
+#' @export
+tdata2env <- function(data) { # nolint
+  checkmate::assert_class(data, "tdata")
+  list2env(lapply(data, function(x) if (is.reactive(x)) x() else x))
+}
+
+#' @rdname tdata
+#' @param x a `tdata` object
+#' @param ... additional arguments for the generic
+#' @export
+get_code.tdata <- function(x, ...) { # nolint
+  # note teal.data which teal depends on defines the get_code method
+  attr(x, "code")()
+}
+
+
+#' Wrapper for `get_code.tdata`
+#' This wrapper is to be used by downstream packages to extract the code of a `tdata` object
+#'
+#' @param data (`tdata`) object
+#'
+#' @return (`character`) code used in the `tdata` object.
+#' @export
+get_code_tdata <- function(data) {
```

Comment:
I don't understand why there are 2 functions here. I could maybe see it if `get_code.tdata` was not exported, but it is. As a user I would be confused as to which one I should use.

Reply:
What about axing `get_code.tdata` and just use `get_code_tdata` then? It feels like an unnecessary complexity to have both.

If we keep `get_code.tdata` though, the generic documentation should be updated. Right now the generic reads `@param x ([`TealDatasetConnector`] or [`TealDataset`]). If of class `character` will be treated as file to read.` It should also say `tdata` in this case.

Reply:
> What about axing get_code.tdata and just use get_code_tdata then? It feels like an unnecessary complexity to have both.

Fair enough